python/ray/data/_internal/dataset_logger.py

```diff
@@ -0,0 +1,130 @@
+import logging
+import os
+
+import ray
+from ray._private.ray_constants import LOGGER_FORMAT
+
+
+class DatasetLogger:
+    """Logger for Ray Datasets which, in addition to logging to stdout,
+    also writes to a separate log file at `DatasetLogger.DEFAULT_DATASET_LOG_PATH`.
+    """
+
+    DEFAULT_DATASET_LOG_PATH = "logs/ray-data.log"
+
+    def __init__(self, log_name: str):
+        """Initialize DatasetLogger for a given `log_name`.
+
+        Args:
+            log_name: Name of logger (usually passed into `logging.getLogger(...)`)
+        """
+        # Primary logger for logging to log file
+        self.logger = logging.getLogger(log_name)
+        # Secondary logger for logging to stdout
+        self.logger_stdout = logging.getLogger(f"{log_name}.stdout")
+        # Clear existing handlers in primary logger
+        # to disable logging to stdout by default
+        while len(self.logger.handlers) > 0:
+            self.logger.removeHandler(self.logger.handlers[0])
+
+        # Explicitly need to set the logging level again;
+        # otherwise the default level is `logging.NOTSET`,
+        # which suppresses logs from being emitted to the file
+        self.logger.setLevel(logging.INFO)
+        self.logger_stdout.setLevel(logging.INFO)
+
+        # Add log handler which writes to a separate Datasets log file
+        # at `DatasetLogger.DEFAULT_DATASET_LOG_PATH`
+        global_node = ray._private.worker._global_node
+        if global_node is not None:
```

Comment:
if `global_node` is not defined, we shall still maintain the existing behavior to able to use `self.logger_stdout`, to log to stdout, and `self.logger` should not do anything.

Reply:
yes, this is the case as in the current implementation (no changes needed to be able to use `self.logger_stdout`). however, the user will need to explicitly use the `log_to_stdout` param when calling `logger.info()`, `logger.warning()`, etc. in order to log to stdout

Reply:
Cool, that makes sense.

Reply:
added a small check to ensure that `self.logger` has valid handlers before calling the individual logging methods, like so:
https://github.com/ray-project/ray/pull/30557/files#diff-3117fe2a5dabf520fb0a2f05380c9ba9b6862421d381f1625e2c4de40ec38616R63-R64